### PR TITLE
feat: Adds room info http endpoint jwt protected.

### PR DESCRIPTION
**Pull request by damencho**
Used from dialplan from jigasi for handling passwords in IVR.

### CODE REVIEW

File: doc/debian/jitsi-meet/jitsi-meet.example [hunk 1]
```diff
@@ -73,6 +73,14 @@ server {
         alias /usr/share/jitsi-meet/libs/external_api.min.js;
     }
 
+    location = /_api_/room-info {
+        proxy_pass http://prosody/room-info?$args;
+        proxy_http_version 1.1;
+        proxy_set_header X-Forwarded-For $remote_addr;
+        proxy_set_header Host $http_host;
+        proxy_set_header Connection "";
```
Comment (saghul) [inline doc/debian/jitsi-meet/jitsi-meet.example:81]: Why do you need this header?

File: resources/prosody-plugins/mod_muc_password_check.lua [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+local room_jid_match_rewrite = module:require "util".room_jid_match_rewrite;
+local starts_with = module:require "util".starts_with;
+local token_util = module:require "token/util".new(module);
+local inspect = require "inspect";
+local parse = require "net.url".parseQuery;
+
+
+-- option to enable/disable room API token verifications
+local enableTokenVerification
+= module:get_option_boolean("enable_password_token_verification", true);
```
Comment (saghul) [inline resources/prosody-plugins/mod_muc_password_check.lua:13]: When would someone want this to be false? If the API is not exposed?

File: resources/prosody-plugins/mod_muc_password_check.lua [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+    return true;
+end
+
+function handle_validate_room_password (event)
+    local request = event.request;
+    module:log("info","Request for room password validate: reqid %s", request.headers["request_id"])
+
+    if request.headers.content_type ~= json_content_type
+            or (not request.body or #request.body == 0) then
+        return 400;
```
Comment (saghul) [inline resources/prosody-plugins/mod_muc_password_check.lua:66]: maybe add a log line?

File: resources/prosody-plugins/mod_muc_password_check.lua [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+    end
+
+    local params = json.decode(event.request.body);
+    if not params then
+        return 400;
```
Comment (saghul) [inline resources/prosody-plugins/mod_muc_password_check.lua:71]: ditto

File: resources/prosody-plugins/mod_muc_password_check.lua [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+    end
+
+    local conference = params["conferenceFullName"];
+    local passcode = params["passcode"];
+
+    if (not conference or not passcode) then
+        return { status_code = 400 };
```
Comment (saghul) [inline resources/prosody-plugins/mod_muc_password_check.lua:78]: why do we return a table here vs just the code in the previous cases?

File: resources/prosody-plugins/mod_muc_password_check.lua [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+        response_data["valid"] = room:get_password() == passcode;
+    else
+        response_data["valid"] = false;
+
+        -- no room found but everything else was valid
+        PUT_response.status_code = 404;
+    end
+
+    PUT_response.body = json.encode(response_data);
+    module:log("info","Sending response for room password validate: %s", inspect(PUT_response));
```
Comment (saghul) [inline resources/prosody-plugins/mod_muc_password_check.lua:114]: shouldn't this be debug level?

**Review by saghul — COMMENTED** [review]
Left some drive-by comments :-P

File: doc/debian/jitsi-meet/jitsi-meet.example (hunk 1 above)
Comment (damencho) [reply doc/debian/jitsi-meet/jitsi-meet.example:81]: Copy paste :)

File: resources/prosody-plugins/mod_muc_password_check.lua (hunk 2 above)
Comment (damencho) [reply resources/prosody-plugins/mod_muc_password_check.lua:13]: Someone may want to use it without the jwt authentication part ... in other virtualhosts and handle exposing it another way ...

File: resources/prosody-plugins/mod_muc_password_check.lua (hunk 5 above)
Comment (damencho) [reply resources/prosody-plugins/mod_muc_password_check.lua:78]: No idea ... will change it. Most of the code was borrowed from another internal module.

File: doc/debian/jitsi-meet/jitsi-meet.example (hunk 1 above)
Comment (damencho) [reply doc/debian/jitsi-meet/jitsi-meet.example:81]: @saghul what about the _api_ , any opinions?

File: resources/prosody-plugins/mod_muc_password_check.lua (hunk 2 above)
Comment (damencho) [reply resources/prosody-plugins/mod_muc_password_check.lua:13]: I will drop it, to simplify the code :)

File: resources/prosody-plugins/mod_muc_password_check.lua (hunk 6 above)
Comment (damencho) [reply resources/prosody-plugins/mod_muc_password_check.lua:114]: I'm changing it ... but I'm hesitating ... maybe we should leave a trace in the logs for a check ...

File: doc/debian/jitsi-meet/jitsi-meet.example (hunk 1 above)
Comment (saghul) [reply doc/debian/jitsi-meet/jitsi-meet.example]: I've see _foo used, but never _foo_.

I wonder if we could have some generic "_service" or "_internal" tenant-like so we put all internal things there like _service/room-info _service/jaas-something.

Then we can filter easy on nginx.

Comment (saghul) [reply doc/debian/jitsi-meet/jitsi-meet.example]: I've see _foo used, but never _foo_.

I wonder if we could have some generic "_service" or "_internal" tenant-like so we put all internal things there like _service/room-info _service/jaas-something.

Then we can filter easy on nginx.

Comment (damencho) [reply doc/debian/jitsi-meet/jitsi-meet.example]: > I wonder if we could have some generic "_service" or "_internal" tenant-like so we put all internal things there like _service/room-info _service/jaas-something.

That was the idea of introducing `_api` ... 

I have a problem with the nginx rule as I need to duplicate it to match

`/_api/room-info` and `/_api/room-info/someusertenant`, but I need to discuss that and with Aaron and Scott as I'm not sure how the haproxy mathcing is working ...

Comment (saghul) [reply doc/debian/jitsi-meet/jitsi-meet.example]: Ah that looks nice! Maybe I just reviewed an old version!

File: doc/debian/jitsi-meet/jitsi-meet.example [hunk 7]
```diff
@@ -73,6 +73,13 @@ server {
         alias /usr/share/jitsi-meet/libs/external_api.min.js;
     }
 
+    location = /_api_/room-info {
```
Comment (saghul) [inline doc/debian/jitsi-meet/jitsi-meet.example:76]: `_api` or `_api_` ? I have seen the former around, but never the latter...

File: resources/prosody-plugins/mod_muc_password_check.lua [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+    end
+    local params = parse(event.request.url.query);
+    local room_name = params["room"];
+    local domain_name = params["domain"];
+    local subdomain = params["subdomain"];
+    local conference = params["conferenceFullName"];
+
+    local room_address;
+
+    if (not conference) and ((not room_name) or (not domain_name)) then
```
Comment (saghul) [inline resources/prosody-plugins/mod_muc_password_check.lua:129]: are the inner parens needed in the right side of the and ?

File: resources/prosody-plugins/mod_muc_password_check.lua [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+
+    if token and starts_with(token,'Bearer ') then
+        token = token:sub(8,#token)
+    end
+
+    if not verify_token(token, room_address) then
+        return 403;
+    end
+
+        local room = get_room_from_jid(room_address);
```
Comment (saghul) [inline resources/prosody-plugins/mod_muc_password_check.lua:156]: Weird indent

File: resources/prosody-plugins/mod_muc_password_check.lua [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+        module:log('debug', 'No host/component found, will wait for it: %s', name)
+
+        -- when a host or component is added
+        prosody.events.add_handler('host-activated', process_host);
+    else
+        process_host(name);
+    end
+end
+
+process_host_module(muc_domain_base, function(host_module, host)
```
Comment (saghul) [inline resources/prosody-plugins/mod_muc_password_check.lua:197]: Would it make sense to add some sort of generic root route for these internal APIs?

File: resources/prosody-plugins/mod_muc_password_check.lua (hunk 5 above)
Comment (damencho) [reply resources/prosody-plugins/mod_muc_password_check.lua]: Actually, the status_code is expected by the handler. 
https://github.com/jitsi/jitsi-meet/blob/ffc412c18d7a6b4dc92e03d8a8a379031d24f4fc/resources/prosody-plugins/util.lib.lua#L149

File: resources/prosody-plugins/mod_muc_password_check.lua [hunk 11]
```diff
@@ -35,145 +35,129 @@ function verify_token(token, room_address)
...
-        response_data["valid"] = room:get_password() == passcode;
+    if not room then
+        module:log("warn", "No room found for %s", room_address);
+        return 404, nil;
     else
-        response_data["valid"] = false;
-
-        -- no room found but everything else was valid
-        PUT_response.status_code = 404;
+        return 200, room;
```
Comment (saghul) [inline resources/prosody-plugins/mod_muc_password_check.lua:95]: What is `room` here, just a string? Or would be returning a complex object?

File: resources/prosody-plugins/mod_muc_password_check.lua [hunk 12]
```diff
@@ -35,145 +35,129 @@ function verify_token(token, room_address)
...
     end
 
+    local PUT_response = {
+        headers = { content_type = "application/json"; };
+        body = json.encode({ valid = (room:get_password() == passcode) })
+    };
 
-    -- verify access
-    local token = event.request.headers["authorization"]
+    module:log("debug","Sending response for room password validate: %s", inspect(PUT_response));
```
Comment (saghul) [inline resources/prosody-plugins/mod_muc_password_check.lua:132]: debug level?

**Review by saghul — APPROVED** [review]
LGTM with a couple of questions.

File: resources/prosody-plugins/mod_muc_password_check.lua (hunk 11 above)
Comment (damencho) [reply resources/prosody-plugins/mod_muc_password_check.lua:95]: It is the room object. An object with methods and such.

File: resources/prosody-plugins/token/util.lib.lua [hunk 13]
```diff
@@ -293,7 +294,7 @@ function Util:verify_room(session, room_address)
     if not self.enableDomainVerification then
         -- if auth_room is missing, this means user is anonymous (no token for
         -- its domain) we let it through, jicofo is verifying creation domain
-        if auth_room and room ~= auth_room and auth_room ~= '*' then
+        if auth_room and (room ~= auth_room and not ends_with(room, ']'..auth_room)) and auth_room ~= '*' then
```
[the remark this answers is not in the document]
Comment (damencho) [reply resources/prosody-plugins/token/util.lib.lua:296]: In most of the places we use tokens with room: "*", and we do not hit the problem. We started specifying the room name for jaas components and voila a bug was revealed :)

[the remark this answers is not in the document]
Comment (saghul) [reply resources/prosody-plugins/token/util.lib.lua:296]: 👍  maybe the docs need updating?

[the remark this answers is not in the document]
Comment (damencho) [reply resources/prosody-plugins/token/util.lib.lua:297]: Just did.

**Review by saghul — APPROVED** [review]

**Pull request merged**